lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php

```diff
@@ -299,8 +299,7 @@ protected function gatherRowData(array $data, array &$id, array &$nonemptyCompon
                     // If there are field name collisions in the child class, then we need
                     // to only hydrate if we are looking at the correct discriminator value
                     if (isset($cacheKeyInfo['discriminatorColumn'], $data[$cacheKeyInfo['discriminatorColumn']])
-                        // Note: loose comparison required. See https://github.com/doctrine/doctrine2/pull/6304#issuecomment-323294442
-                        && ! in_array($data[$cacheKeyInfo['discriminatorColumn']], $cacheKeyInfo['discriminatorValues'])
+                        && ! in_array((string) $data[$cacheKeyInfo['discriminatorColumn']], $cacheKeyInfo['discriminatorValues'], true)
```

Comment:
In that comment @Ocramius says:

> I checked the issue locally, and it's because the discriminator is not converted to a PHP value.
> 
> That said, trying to fix the issue leads to a massive can of worms that is being opened: let's keep the loose comparison there for now.

I've opened the can of worms and it wasn't that bad but it's possible that I've overlooked things, please use some ❤️ to review this part